### DISCUSSION TITLE
is there a good setup for an emacs user that does not switch between insert and normal modes?

### DISCUSSION THREAD

**Discussion by hoebat** [Olical/conjure#226] [Q&A]
Hi @Olical 

I've been playing around with conjure's features and it's fantastic. However, coming from emacs, I'm having issues just moving around in the editor. I don't know how to switch files or how to switch buffers in vim so it's basically unusable for me until I figure out how to get it working.

The hjkl thing is not really that great on a Dvorak keyword so I don't intend do jump through those hoops. I also really can't get use to pressing `I` for insert to type anything because I spend a lot more time moving around and it interrupts the flow too much. I have a couple of keybindings that help me with moving/editing that I'd like to get working in Vim:

- `C-<right>` and `C-<left>` map to slurp and spit 
- `C-e` is eval last sexp and `C-x-x` is eval file
- `C-<up>` and `C-<down>` is for moving up and down forms
- `C-x-<right>` and `C-x-<left>` to switch back and forth between buffers

I also use M-. and M-, for jumping to and back from definitions. Anything else is optional for me.

Is there some sort of `emacs` mode for vim or a sample emacs-ish config that I can use to start myself off with conjure?

**Comment by Olical** [comment]
Moved to a discussion since it's more of a generic free form vim question than a Conjure one.

I don't have any direct advice to help you down the current path of "insert first" vim usage, my advice would be to embrace normal mode and the modal nature of vim. I spend a fraction of my time in insert mode actually typing and 95% in normal mode, hopping around and issuing commands such as `yi)` to "yank inside matching parenthesis" then `j10$bp` to "go down 10 lines, go to the end of the line, go back a word and paste what I yanked".

Vim's motions, text objects and mappings are incredibly powerful, I can't edit for any length of time without them, I feel so held back by moving a cursor character by character using arrow keys and backspace. Things like [easymotion](https://github.com/easymotion/vim-easymotion) take that power further by allowing you to project your movements anywhere on your screen instantly without touching your mouse.

So, I'll share this on twitter and see if others have advice but my recommendation is to embrace the evil side. I tried emacs+evil mode for over a year (Spacemacs) but ran into too many siturations where the vim mappings fell away and I was left with unfamiliar emacs mappings. My solution was to go back to vim and write Conjure to scratch the eval itch.

**Reply by hoebat** [reply]
Yeah. I get it but it's hard to break some habits.

I tried to get used to vim (about 10 years ago) but once I started using Dvorak, I realised that most of the inefficiencies are through typing and having the hand go through a bunch of awkward positions for simple words like `the` and `for`. 

I'll go through my rational (though it's not very likely to convince anyone significantly invested in the vim workflow): 
- The keyboard on a laptop is pretty small and the arrow keys are not that far away. 
- The touchpad is also super close on the laptop and it's the best for smallish fine grained cursor selection
- The `mode change` occurs when I move my hands to the arrow keys or I scroll with the touchpad and I don't mind tapping a couple of times interspersed used with `ctrl` (for fast movement) because it's a relatively braindead operation that doesn't break flow.
- I'll emphasise the `Braindead` movement part.  I can use most of my brain for thinking about what I'm  going to do next as opposed to what characters I'm going to type in order to get me to the next position. 
- I rely on `swiper` https://github.com/abo-abo/swiper - mapped to `C-l` for locating and jumping to terms. I also use `goto line` a lot but not as often so that gets mapped to `C-x-g` (two keypresses)
-  I do a lot of eval/micro adjustments  and wish to do it with minimal keystrokes:
   1. Move to a spot, type some characters, select out the autocomplete and then adjust the parens with spit/slurp.
   2. Type some characters, select autocomplete and immediately eval
- I have slurp and spit mapped to `C-<right>` and `C-<left>` because it has to be fast and does not break flow. Similarly I map `C-<e>` instead of `C-x-e` because it's such a common thing to do. Then typing, completing and evaluating is just one or two keystrokes past the actual characters I type.
- I can't do that switching between insert and normal modes because that already takes 2 keystrokes every mode switch and back (not to mention that hand position change to `Esc` when switching).

----

Anyways, besides that I have to say that the Neovim tooling for clojure is starting to get really good. I do miss `Elisp` and wish that  I can open Neovim and start writing and evaling a lisp like `fennel` straight away in a scratch buffer. That and the no mode switch thing.

I took a look at Neovim's api and it's fantastic so theoretically, it is possible to write an emacs-like front end for it but yeah... it's a lot of work.

**Comment by daveyarwood** [comment]
FWIW, I learned Dvorak years before Vim, and when I started using Vim, I was also a little concerned that hjkl wouldn't be convenient, but I quickly found that (at least, in my opinion!) hjkl still feels pretty good, even in the Dvorak layout:

* h (left) and l (right)  fall under your right pointer and pinky, i.e. your leftmost and rightmost finger, which feels intuitive enough
* j (down) and k (up) fall under the left middle finger and pointer, so if I want to go up and down, I just move those two fingers down a row and it also feels pretty intuitive.

It's also worth noting that as you get more comfortable with Vim, if you're "doing it right," you won't be using hjkl to get around anyway, most of the time.

My suggestion would be to give vimtutor a try, spend a few days with Vim trying to do things the Vim way, and see what you think. You might be surprised to find that you enjoy it! (Or maybe not! YMMV :)  )

**Comment by cvkline** [comment]
I agree with Oliver that if one is to embrace vi-style editors, one really must embrace the modes (command, insert, visual). I get that on a Dvorak keyboard the hjkl movement keys are in somewhat awkward places, but the arrow keys work just as well in that case, and you're already using them in Emacs. As I said elsewhere, vi's modes are a learning curve, but I believe a very worthwhile one.

**Comment by hoebat** [comment]
Look. I'm not getting into dogma. If we're purely talking about ergonomics, the emacs keybindings out of the box are not that great either. I got so frustrated with the emacs keybindings that I wrote my own rebinding library.

https://github.com/zcaudate/eta

and bound most of the commonly used functions to `<F1>`, `<F2>` ... etc.  It's the only way I can remember keys.

Here are my bindings for reference.

https://github.com/zcaudate/etude/blob/master/etude/etude-bindings.el

I've also installed a keylogger in emacs and looked at how I'm using the editor.

https://github.com/zcaudate/eta-logger

I've got around 7 months of data confirming that I spend most of my time on either movement or eval, even when I'm heavily writing code because about 80% of the time, I'm actually evaling on a test example to make sure that the outputs are correct.

Based on my usage I believe that mnemonic movement and modes actually will actually increase my keystrokes and decrease my productivity overall. Unless there's actual data that show a significant increase in productivity for mnemonic movement and modes, it's just personal experience. I'm not discounting personal experience but I have my own ones as well and these days, I don't get swayed too easily unless someone shows me hard evidence.

---

I know that vim works for a lot of people. It's great. NeoVim is something that I'll most likely get into very slowly because luajit rocks and the design is so much simpler than emacs. If I need to write programs for vim - I'll most likely do it by connecting to it via the msgpack api (through Clojure and emacs) and sending it lua commands. It'd be faster for me to write programs for nvim that way than writing those programs in vim itself.
